ui/src/server_axum.rs

```diff
@@ -0,0 +1,511 @@
+        .route("/evaluate.json", post(evaluate))
+        .route("/compile", post(compile))
+        .route("/execute", post(execute))
+        .route("/format", post(format))
+        .route("/clippy", post(clippy))
+        .route("/miri", post(miri))
+        .route("/macro-expansion", post(macro_expansion))
+        .route("/meta/crates", get_or_post(meta_crates))
+        .route("/meta/version/stable", get_or_post(meta_version_stable))
+        .route("/meta/version/beta", get_or_post(meta_version_beta))
+        .route("/meta/version/nightly", get_or_post(meta_version_nightly))
+        .route("/meta/version/rustfmt", get_or_post(meta_version_rustfmt))
+        .route("/meta/version/clippy", get_or_post(meta_version_clippy))
+        .route("/meta/version/miri", get_or_post(meta_version_miri))
+        .route("/meta/gist/", post(meta_gist_create))
+        .route("/meta/gist/:id", get(meta_gist_get))
+        .route("/metrics", get(metrics))
+        .layer(AddExtensionLayer::new(Arc::new(SandboxCache::default())))
+        .layer(AddExtensionLayer::new(config.github_token()));
+
+    if let Some(token) = config.metrics_token() {
+        app = app.layer(AddExtensionLayer::new(token))
+    }
+
+    if config.use_cors() {
+        app = app.layer({
+            CorsLayer::new()
+                .allow_origin(cors::any())
+                .allow_headers([header::CONTENT_TYPE])
+                .allow_methods([Method::GET, Method::POST])
+                .allow_credentials(false)
+                .max_age(ONE_HOUR)
+        });
+    }
+
+    // Basic access logging
+    app = app.layer(TraceLayer::new_for_http());
+
+    axum::Server::bind(&config.server_socket_addr())
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+fn get_or_post<T: 'static>(handler: impl Handler<T> + Copy) -> MethodRouter {
+    get(handler).post(handler)
+}
+
+fn static_file_service(root: impl AsRef<path::Path>, max_age: HeaderValue) -> MethodRouter {
+    let files = ServeDir::new(root).precompressed_gzip();
+
+    let with_caching = SetResponseHeader::if_not_present(files, header::CACHE_CONTROL, max_age);
+
+    get_service(with_caching).handle_error(|e| async move {
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            format!("Unhandled internal error: {}", e),
+        )
+    })
+}
+
+async fn rewrite_help_as_index<B>(
+    mut req: Request<B>,
+    next: middleware::Next<B>,
+) -> impl IntoResponse {
+    let uri = req.uri_mut();
+    if uri.path() == "/help" {
+        let rewritten_uri = mem::take(uri);
+        let mut parts = rewritten_uri.into_parts();
+        parts.path_and_query = Some(PathAndQuery::from_static("/index.html"));
+        *uri = Uri::from_parts(parts).unwrap();
+    }
+    next.run(req).await
+}
+
+// This is a backwards compatibilty shim. The Rust documentation uses
+// this to run code in place.
+async fn evaluate(Json(req): Json<EvaluateRequest>) -> Result<Json<EvaluateResponse>> {
```

Comment:
Do you care about the response sent back if the `Json` extractor rejects the request? [These](https://docs.rs/axum/latest/axum/extract/rejection/enum.JsonRejection.html) are the reasons that can happen. It'll return a plain text response in that case.

Customizing these "rejection responses" is done [like this](https://docs.rs/axum/latest/axum/extract/index.html#customizing-extractor-responses). Its a bit more boilerplate that I'd like but its the best solution we have thus far.

Reply:
Hmm. I'll have to investigate what the current behavior is. IIRC, I recover and print out a JSON-formatted error.